p2p/src/peer_manager/peerdb/mod.rs

```diff
@@ -0,0 +1,258 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Peer database
+//!
+//! The peer database stores:
+//! - connected peers
+//! - available (discovered) addresses
+//! - banned addresses
+//!
+//! Connected peers are those peers that the [`crate::peer_manager::PeerManager`] has an active
+//! connection with. Available addresses are discovered through various peer discovery mechanisms and they are
+//! used by [`crate::peer_manager::PeerManager::heartbeat()`] to establish new outbound connections
+//! if the actual number of active connection is less than the desired number of connections.
+
+use std::{
+    collections::{BTreeMap, HashMap, HashSet},
+    sync::Arc,
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+
+use logging::log;
+
+use crate::{
+    config,
+    error::{ConversionError, P2pError},
+    interface::types::ConnectedPeer,
+    net::{
+        types::{self, Role},
+        AsBannableAddress, NetworkingService,
+    },
+};
+
+#[derive(Debug)]
+pub struct PeerContext<T: NetworkingService> {
+    /// Peer information
+    pub info: types::PeerInfo<T>,
+
+    /// Peer's address
+    pub address: T::Address,
+
+    /// Peer's role (inbound or outbound)
+    pub role: Role,
+
+    /// Peer score
+    pub score: u32,
+}
+
+impl<T: NetworkingService> From<&PeerContext<T>> for ConnectedPeer {
```

Comment:
Why convert from ref?

Reply:
And then clone strings?

It's always preferable to delegate the clone decision to the developer using the feature.

Reply:
> Why convert from ref?

Because we convert from ref here:

```
        self.peers.values().map(Into::into).collect()
```

Converting from value will require implement `Clone` for `PeerContext`.
And I don't think it's a good idea:

- `PeerContext` probably should not be cloneable for safety.
- Cloning `PeerContext` will make a clones of some non-trivial types (two `BTreeSet` for example) that will be dropped once we convert to `ConnectedPeer`.
- We might want to add some other types to `PeerContext` someday that do not implement `Clone`.

So I think converting from ref is better.

Reply:
I'm not really convinced, but I'm gonna let this go for now because it doesn't really matter and maybe I just don't have the full picture.